LoopKit/DailyValueSchedule.swift

```
@@ -194,25 +165,97 @@ public struct DailyValueSchedule<T: RawRepresentable>: RawRepresentable, CustomD
+            switch (nextLeftStartTime, nextRightStartTime) {
+            case (.some(let leftStart), .some(let rightStart)):
+                if leftStart < rightStart {
+                    leftCursor += 1
+                } else if rightStart < leftStart {
+                    rightCursor += 1
+                } else {
+                    leftCursor += 1
+                    rightCursor += 1
+                }
+            case (.some, .none):
+                leftCursor += 1
+            case (.none, .some):
+                rightCursor += 1
+            case (.none, .none):
+                leftCursor += 1
+                rightCursor += 1
+            }
+        } while leftCursor < lhs.items.endIndex && rightCursor < rhs.items.endIndex
+
+        return DailyValueSchedule(dailyItems: alignedItems, timeZone: lhs.timeZone)!
+    }
 }
 
 
-extension DailyValueSchedule where T: Equatable {
-    public static func ==(lhs: DailyValueSchedule<T>, rhs: DailyValueSchedule<T>) -> Bool {
-        guard lhs.items.count == rhs.items.count else {
-            return false
+extension DailyValueSchedule: RawRepresentable, CustomDebugStringConvertible where T: RawRepresentable {
+    public typealias RawValue = [String: Any]
+    public init?(rawValue: RawValue) {
+        guard let rawItems = rawValue["items"] as? [RepeatingScheduleValue<T>.RawValue] else {
+            return nil
         }
 
-        guard lhs.timeZone == rhs.timeZone else {
-            return false
-        }
+        var timeZone: TimeZone?
 
-        for (a, b) in zip(lhs.items, rhs.items) {
-            guard a == b else {
-                return false
-            }
+        if let offset = rawValue["timeZone"] as? Int {
+            timeZone = TimeZone(secondsFromGMT: offset)
         }
 
-        return true
+        self.init(dailyItems: rawItems.compactMap { RepeatingScheduleValue(rawValue: $0) }, timeZone: timeZone)
```

Technically, the `compactMap` here hides failures. That is, if a `RepeatingScheduleValue` failed to init from its rawValue, the overall `self.init` here would succeed (due to the `compactMap` eliminating any `nil`s, but the resulting `DailyValueSchedule` would not be the same as what was serialized. It technically shouldn't happen, *unless* there was a bug or backwards incompatible update.

---

You're right; what would you expect the behavior to be in the case of a `nil`?

---

I would expect any failure from a contained `RepeatingScheduleValue` to also fail this `init?` (so it would also return `nil`).

---

That makes sense; I'll make the change here and at the other `compactMap` sites.